src/vm/vm_memory/memory.rs

```diff
@@ -149,7 +224,7 @@ impl Memory {
             return Err(MemoryError::NonZeroOffset(src_ptr.offset));
         }
 
-        let segment_index = -src_ptr.segment_index as usize;
+        let segment_index = -(src_ptr.segment_index + 1) as usize;
```

Comment:
Why did this change?

Reply:
Before, the first temporary segment (-1) had key -1, not zero, which is unintuitive and may cause problems in the extreme case that there are `isize::MIN` temporary segments.

Reply:
I would also recommend changing the test` relocate_value_mayberelocatable_temporary_segment_rules`, or creating a parallel test that uses add_relocation_rule() to add rules as opposed to initializing it directly

Reply:
Regarding my first reply, I also think that having this extra logic makes relocation rules more confusing, and what we gain from it (being able to host an extra relocation_rule) doesn't quite make up for the rise in complexity (shifting relocation rules indexes to the left)

Reply:
The snippet you quoted is from `add_relocation_rule()`, so it's already reflected in that method. Of course, it's also reflected in the added method (`Memory::relocate_memory()`). The idea is that this change is only about the internal representation and shouldn't affect anything outside `Memory`.

I completely agree about the test case. I've fixed it already.

My reasoning is that by considering the normal and temporary segments as separate lists that begin with index zero, it wouldn't make sense to have it begin at one at `HashMap` where the key contains exclusively temporary segment indices. In other words, I think it's better to keep it zero-based because `temp_data` is a zero-based list, and the key links to that structure's indices. If it's still confusing I can explicitly document it with a comment or revert it as it was before.

Reply:
Adding comments  on all affected methods should be enough to prevent future confusion

Reply:
Done. I've also made them private (`relocation_rules` and `validation_rules`) since they are neither accessed nor meant to be from the outside. This way the change is kept strictly internal.